### ISSUE TITLE
How to use ApiPromise gently?

### ISSUE THREAD

**Issue by clearloop** [polkadot-js/api#2156]
Our project has lots of logic needs to be sync, but I find that calling `await api.query.xxxx` can not block substrate request completely.

## For example
 
```javascript
async function parseRes(r: any) {
    const status = r.status;
    log(`Transaction status: ${status.type}`);

    log(status);
    if (status.isInBlock) {
        log(`Included at block hash: ${status.asInBlock.toHex()}`);
        r.events && r.events.forEach(async (r: any) => {
            log(
                "\t" +
                r.phase.toString() +
                `: ${r.event.section}.${r.event.method}` +
                r.event.data.toString(),
            );

            if (r.event.data[0].isModule) {
                const doc = await this.api.registry.findMetaError(r.event.data[0].asModule);
                const err = `${doc.name}.${doc.section} - ${doc.documentation.join(" ")}`;
                log(err, Logger.Error);
            }
        });
    } else if (status.isFinalized) {
        log(`Finalized block hash: ${status.asFinalized.toHex()}`);
        this.queue.active = false;
    }
}
```

We need to block the `parseRes` till `status.isFinalized`, is there any way to reach this?

## Ugly solution

I had to make 3 intervals in my program to check the status and make the process sync...and this is really a bad solution...

**Comment by jacogr** [comment]
1. use `new Promise((resolve, reject) => {... })`
2. on `isFinalized` (i.e. 3 blocks behind inclusion), `resolve`, on error `reject`
3. wait for that promise

**Comment by clearloop** [comment]
> 1. use `new Promise((resolve, reject) => {... })`
> 2. on `isFinalized` (i.e. 3 blocks behind inclusion), `resolve`, on error `reject`
> 3. wait for that promise

Well I know, just do not use `async / await`? yep? I'll try this and close this issue after I succeed!

**Comment by jacogr** [comment]
With await, what happens if finalization is stalled? The promise will never resolve.

The real reason is API consistency. these are subscriptions, where events stream in, there is not just a single update, rather updates are sent while the tx is processed. 

Everywhere where callbacks are passed, the API always returns the unsubscription Promise. So if we have event subscriptions here, without unsub but rather an end-result, we are completely out-of-sync with how everything is done. So everywhere -

- await with no callback - resolves with the first result (for extrinsics this is the hash as per the non-sub RPC)
- await with callback - resolves with the unsub function, the user needs to determine when it is the right time to unsubscribe

Additionally we will have issues the base assumptions don't match other assumptions - for instance, in my specific use-cases, unlike yours, I never rely on isFinalized, only when inBlock is not received. (isFinalized is handled as a second tier where it comes in alongside/before InBlock)

**Comment by clearloop** [comment]
**"I don't know Javascript"** real-world version...

---

Thanks for your explanation! You just saved my ass for the twice time!! I got a plan to rebuild our project now~

**Comment by jacogr** [comment]
So there is def a need and requests for "auto-unsub". So what I have been pondering is possibly (optionally) passing an options struct after the callback to switch on this type of behavior, alongside the type of statuses you would want. (With some sane defaults)

It is still damn tricky though, fo instance the `Invalid` status should mean that the tx won't be processed at all and it will be rejected - in the real world (out there currently) it means that it is invalid on the RPC node, but it still be included elsewhere. (So normally you would see a Retracted, then Invalid... and then Finalized...)

So watch this space... there may be some extra options that will allow you to tweak a bit (off by default as to keep consistency), but you certainly are not the first one asking for it and won't be the last. 

TL;DR I do appologize for the extra hoops in this case, I hear the frustration, whatever comes in shouldn't break user-space (keep backwards compat) and been solid, which _may_ be tricky.

**Comment by clearloop** [comment]
> So there is def a need and requests for "auto-unsub". So what I have been pondering is possibly (optionally) passing an options struct after the callback to switch on this type of behavior, alongside the type of statuses you would want. (With some sane defaults)

I like this idea, most of the new developers of substrate used to use `web3.js` I think, folks who have been accustomed to blocking tx till they are finalized.

I used to complain that `polkadot/api` has an ugly design...now I feel sorry about it XD

## InBlock

The `InBlock` idea is reasonable, and it is part of `polkadot/apps` I think?  projects like `polkadot/apps` really need this, but me as a script developer, I'd like the `isFinalized` idea better...because it is easy to use~

## web3

I think `polkadot/api` can draw some design from web3 :)

```js
async sendTx() {                                                                                                                                                                                            
    const addr = this.web3.eth.accounts.wallet[0].address;                                                                                                                                                  
    const contract = burn(this.web3, addr);                                                                                                                                                                 
                                                                                                                                                                                                            
    log("this might take a long time, you can walk around and take a cup of coffee...");                                                                                                                    
    log("do not try to use proxy if you are testing a local darwinia node : )");                                                                                                                            
                                                                                                                                                                                                            
    return await contract.send({                                                                                                                                                                            
        from: addr,                                                                                                                                                                                         
        gas: 500000,                                                                                                                                                                                        
    }).on("transactionHash", (hash: string) => {                                                                                                                                                            
        log(`our tx hash is ${hash}`);                                                                                                                                                                      
    }).on("receipt", (r: any) => {                                                                                                                                                                          
        this.headers.receiptHash = r.transactionHash;                                                                                                                                                       
        this.queue.active = false;                                                                                                                                                                          
    }).on("error", () => {                                                                                                                                                                                  
        log([                                                                                                                                                                                               
            "send redeem request to eth contract failed, please make sure ",                                                                                                                                
            "you have enough ether and ring in your ethereum account. if you",                                                                                                                              
            "haven't change the default eth account, check the default one: ",                                                                                                                              
            "0x4c0883a69102937d6231471b5dbb6204fe5129617082792ae468d01a3f362318 ",                                                                                                                          
            "if it has any eth and ring left.\n\n",                                                                                                                                                         
            "if you think this is a bug, please raise an issue at: ",                                                                                                                                       
            "https://github.com/darwinia-network/darwinia-offchain-tools/issues/new"                                                                                                                        
        ].join(""), Logger.Error);                                                                                                                                                                          
    });                                                                                                                                                                                                     
}                 
```

I love this `.on` design : D

**Comment by clearloop** [comment]
Btw, have you ever think about writing `polkadot/api` in wasm? I want to give it a try!

If `parity` have some plan about it, let me know and I'll pr it till it gets perfect~

**Comment by jacogr** [comment]
The `InBlock` vs `IsFinalized` is funny - not all chains support both (yet). The older chains have `IsFinalized` meaning the new `InBlock`.

Wasm is interesting, although the stuff like the types are going to create some issues. For instance all the teams replicating this functionality even in Rust, have some issues along the way. Immediately no plans in that direction, although some more stuff may probably move into WASM, like we have the crypto/hashing in there.

the other issue however with is a bit of nightmare, is the React Native support for wasm - there are some serious hoops to jump through there (effectivelt `@polkadot/wasm-crypto` also outputs to asm.js for those environments)

**Comment by clearloop** [comment]
> a bit of nightmare, is the **React Native support for wasm**

Got it, I'm persuaded : D
